spateo/digitization/contour.py

```diff
@@ -0,0 +1,276 @@
+    cluster_ids = [str(c) for c in cluster_ids]
+
+    u, count = np.unique(adata_low_res.obs["cluster_img_label"], return_counts=True)
+    count_sort_ind = np.argsort(-count)
+    cluster_labels = u[count_sort_ind]
+    cluster_labels = [c for c in cluster_labels]
+
+    adata_high_res.obs[domain_key] = "NA"
+
+    for i in range(len(cluster_ids)):
+        ctrs, _, _ = extract_cluster_contours(
+            cluster_label_image, cluster_labels[i], bin_size=bin_size_low, k_size=k_size, min_area=min_area, show=False
+        )
+        for j in range(len(adata_high_res)):
+            x = adata_high_res.obsm[spatial_key][j, 0]
+            y = adata_high_res.obsm[spatial_key][j, 1]
+            for k in range(len(ctrs)):
+                if cv2.pointPolygonTest(ctrs[k], (y, x), False) >= 0:
+                    adata_high_res.obs[domain_key][j] = cluster_ids[i]
+
+
+def gen_contour_img(
+    adata: AnnData,
+    bin_size: int = None,
+    spatial_key: str = "spatial",
+    label_key: str = "cluster_img_label",
+    show: bool = True,
+    save_fig: str = "plot_contour_img",
+) -> None:
+    """Generate an image with contours of each spatial domains.
+
+    Args:
+        adata (AnnData): The adata object used to create the image.
+        bin_size (int, optional): The size of the binning. Defaults to None.
+        spatial_key (str, optional): The key name of the spatial coordinates. Defaults to "spatial".
+        label_key (str, optional): The key name of the image label values. Defaults to "cluster_img_label".
+        show (bool, optional): Visualize the result. Defaults to True.
+        save_fig (str, optional): Save image to path or filename. Defaults to "plot_contour_img".
+    """
+
+    import matplotlib.pyplot as plt
+    from numpngw import write_png
+
+    label_list = np.unique(adata.obs[label_key])
+    labels = np.zeros(len(adata))
+    for i in range(len(label_list)):
+        labels[adata.obs[label_key] == label_list[i]] = i + 1
+
+    label_img = np.zeros(
+        (
+            int(max(adata.obsm[spatial_key][:, 0] // bin_size)) + 1,
+            int(max(adata.obsm[spatial_key][:, 1] // bin_size)) + 1,
+        )
+    )
+    for i in range(len(adata)):
+        label_img[
+            int(adata.obsm[spatial_key][i, 0] // bin_size), int(adata.obsm[spatial_key][i, 1] // bin_size)
+        ] = labels[i]
+
+    contour_img = label_img.copy()
+    contour_img[:, :] = 255
+    for i in np.unique(label_img):
+        if i == 0:
+            continue
+        label_img_gray = np.where(label_img == i, 0, 1).astype("uint8")
+        _, thresh = cv2.threshold(label_img_gray, 0, 255, cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)
+        contour, _ = cv2.findContours(thresh, cv2.RETR_CCOMP, cv2.CHAIN_APPROX_NONE)
+        contour_img = cv2.drawContours(contour_img, contour[:], -1, 0.5, 1)
+    if show:
+        plt.imshow(contour_img, cmap="Blues")
+    write_png(save_fig + ".png", contour_img.astype("uint8"))
```

Comment:
Any more decent ways to save an image to file?

Reply:
we have the save figure function. I will look into this